misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,216 @@
+    summary['IP addresses'] = len(df_apiruns['client_ip'].unique())
+    summary['Errors'] = df_w_groups_filtered[df_w_groups_filtered['result'].str.lower().str.contains('error')][
+        'api_call_group_id'].nunique()
+
+    # Value set combos
+    df_get_csets = df_w_groups_filtered[df_w_groups_filtered['api_call'] == 'get-csets']
+    summary['Value set combos'] = len(set([tuple(x) for x in df_get_csets['codeset_ids']]))
+
+    # todo: temp: analysis
+    #  - analyze uniqueness of get-csets calls within an API call group session
+    # diff_get_csets_vs_groups=35; i feel like this should be equal to some calc involving next 4 vars, but not sure
+    # noinspection PyUnusedLocal
+    diff_get_csets_vs_groups = len(df_get_csets) - len(df_get_csets['api_call_group_id'].unique())  # 35
+    diff_codeset_ids_in_group__n_instances = 0  # 1
+    diff_codeset_ids_in_group__n_calls = 0  # 2
+    mult_get_cset_in_session__n_instances = 0  # 30
+    mult_get_cset_in_session__n_calls = 0  # 64
+    # todo: maybe convert to str in pre-processing instead, if need 'str' more than temporarily
+    df_get_csets['codeset_ids'] = df_get_csets['codeset_ids'].astype(str)
+    for group_id, group_data in df_get_csets.groupby('api_call_group_id'):
+        if len(group_data) > 1 and len(group_data['codeset_ids'].unique()) > 1:
+            diff_codeset_ids_in_group__n_instances += 1
+            diff_codeset_ids_in_group__n_calls += len(group_data)
+        elif len(group_data) > 1:
+            mult_get_cset_in_session__n_instances += 1
+            mult_get_cset_in_session__n_calls += len(group_data)
+
+    df = pd.DataFrame([{'Measure': k, 'Value': v} for k, v in summary.items()])
+    if outpath:
+        df.to_csv(outpath, index=False)
+    return df
+
+
+def run(use_cache=False, verbose=False):
+    """Run analysis
+
+    :param use_cache: If True, will use most recent local CSV instead of calling the database"""
+    # Initial setup ---
+    setup()
+    df_apiruns: pd.DataFrame = get_dataset_with_mods(api_runs_query, USAGE_UNJOINED_CSV_PATH, use_cache, verbose)
+    df_w_groups_filtered: pd.DataFrame = get_dataset_with_mods(usage_query, USAGE_JOINED_CSV_PATH, use_cache, verbose)
+    df_apiruns_dev0: pd.DataFrame = preprocess_dev_ips(df_apiruns, verbose)
+    df_w_groups_filtered_dev0: pd.DataFrame = preprocess_dev_ips(df_w_groups_filtered, verbose)
+
+    # Table ---
+    # Stats: With dev IPs included
+    df_out_dev1: pd.DataFrame = summary_stats(df_apiruns, df_w_groups_filtered)
+
+    # Stats: With def IPs filtered out
+    df_out_dev0: pd.DataFrame = summary_stats(df_apiruns_dev0, df_w_groups_filtered_dev0)
+
+    # Join different output datasets
+    df_out = df_out_dev1.merge(df_out_dev0.rename(columns={'Value': 'Value_no_dev'}), on='Measure', how='outer')
+    df_out.to_csv(OUT_CSV_PATH_ALL, index=False)
+
+    # Plots ---
+    # todo: consider dev + non-dev in same plot
+    # Histogram: Value set n selections
```

Comment:
## Histogram: Value set n selections

![code_set_comparison_size](https://github.com/jhu-bids/TermHub/assets/13045020/5a978b1a-bffe-4001-8c31-b4e248ca5852)


![code_set_comparison_size_-_dev_data_included](https://github.com/jhu-bids/TermHub/assets/13045020/3f8edb84-13eb-4776-b9fa-27324bf5c9df)


I was surprised to see that the distribution was very similar. Although I'm not comparing dev to non-dev. Just dev+non-dev to non-dev. Looking at dev alone would probably be different.

Also, if we check w/out filtering the -1 `api_call_group_id`, the max comparison is bigger:
```py
df_apiruns['len_codeset_ids'] = df_apiruns['codeset_ids'].apply(lambda x: len(x))
print(max(df_apiruns['len_codeset_ids']))
>>> 20
```

I think these numbers are small enough that scaling / log scaling is not necessary.